src/compiler/checker.ts

```diff
@@ -34196,7 +34196,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // that include {} (we know that the other types in such intersections are assignable to object
             // since we already checked for that).
             if (hasEmptyObjectIntersection(rightType)) {
-                error(right, Diagnostics.Type_0_may_represent_a_primitive_value_which_is_not_permitted_as_the_right_operand_of_the_in_operator, typeToString(rightType));
+                const constraint = getBaseConstraintOfType(rightType);
+                if (!constraint || isEmptyAnonymousObjectType(constraint)) {
+                    error(right, Diagnostics.Type_0_may_represent_a_primitive_value_which_is_not_permitted_as_the_right_operand_of_the_in_operator, typeToString(rightType));
+                }

```

Comment:
Instead of adding these lines of code, I'd suggest modifying `hasEmptyObjectIntersection` to

```ts
    function hasEmptyObjectIntersection(type: Type): boolean {
        return someType(type, t => t === unknownEmptyObjectType || !!(t.flags & TypeFlags.Intersection) && isEmptyAnonymousObjectType(getBaseConstraintOrType(t)));
    }
```

Accomplishes the same thing with less code.

Reply:
Nice, I've pushed out the requested change.